rust/worker/src/execution/operator.rs

```diff
@@ -66,18 +100,58 @@ pub(crate) trait TaskWrapper: Send + Debug {
 #[async_trait]
 impl<Input, Output, Error> TaskWrapper for Task<Input, Output, Error>
 where
-    Error: Debug,
+    Error: Debug + Send,
     Input: Send + Sync + Debug,
     Output: Send + Sync + Debug,
 {
     async fn run(&self) {
-        let result = self.operator.run(&self.input).await;
-        let task_result = TaskResult {
-            result,
-            task_id: self.task_id,
+        let result = AssertUnwindSafe(self.operator.run(&self.input))
+            .catch_unwind()
+            .await;
+
+        match result {
+            Ok(result) => {
+                // If this (or similarily, the .send() below) errors, it means the receiver was dropped.
+                // There are valid reasons for this to happen (e.g. the component was stopped) so we ignore the error.
+                let _ = self
+                    .reply_channel
+                    .send(
+                        TaskResult {
+                            result: result.map_err(|e| TaskError::TaskFailed(e)),
+                            task_id: self.task_id,
+                        },
+                        None,
+                    )
+                    .await;
+            }
+            Err(panic_value) => {
+                #[allow(clippy::manual_map)]
+                let panic_value = if let Some(s) = panic_value.downcast_ref::<&str>() {
+                    Some(&**s)
+                } else if let Some(s) = panic_value.downcast_ref::<String>() {
+                    Some(s.as_str())
+                } else {
+                    None
+                };
+
+                let _ = self
+                    .reply_channel
+                    .send(
+                        TaskResult {
+                            result: Err(TaskError::Panic(panic_value.map(|s| s.to_string()))),
+                            task_id: self.task_id,
+                        },
+                        None,
+                    )
+                    .await;
+
+                // Re-panic so the message handler can catch it
+                panic!(
```

Comment:
a little weird but I think this makes sense, since the panic technically happened within a component